drivers/clock_control/clock_control_it8xxx2.c

```diff
@@ -0,0 +1,337 @@
+ */
+#define DT_DRV_COMPAT ite_it8xxx2_ecpm
+
+#include <soc.h>
+#include <zephyr/logging/log.h>
+#include "soc_espi.h"
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/clock/it8xxx2_clock.h>
+#include <zephyr/drivers/clock_control/it8xxx2_clock_control.h>
+
+LOG_MODULE_REGISTER(clock_control_it8xxx2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define PLLFREQ_MASK BIT_MASK(4)
+#define MCUFREQ_MASK BIT_MASK(3)
+#define EC_FREQ_MASK BIT_MASK(4)
+#define FNDFREQ_MASK GENMASK(7, 4)
+
+struct clock_control_it8xxx2_config {
+	struct ecpm_it8xxx2_regs *reg_base;
+	int pll_freq;
+	uint32_t cpu_clock;
+};
+
+const struct clock_control_it8xxx2_config *clock_config;
+
+/* ITE IT8XXX2 ECPM Functions */
+static inline int it8xxx2_clock_control_on(const struct device *dev,
+					clock_control_subsys_t sub_system)
+{
+	const struct clock_control_it8xxx2_config *cfg = dev->config;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = cfg->reg_base;
+	struct it8xxx2_clock_control_cells *clk_cfg =
+		(struct it8xxx2_clock_control_cells *)(sub_system);
+
+	uint8_t offset = clk_cfg->offset;
+	uint8_t code = clk_cfg->bitmask;
+	uint8_t always = clk_cfg->always;
+
+	volatile uint8_t *reg =
+		(volatile uint8_t *)((uint32_t)ecpm_regs + (uint32_t)offset);
+
+	*reg = (*reg & ~code) | always;
+
+	return 0;
+}
+
+static inline int it8xxx2_clock_control_off(const struct device *dev,
+					clock_control_subsys_t sub_system)
+{
+	const struct clock_control_it8xxx2_config *cfg = dev->config;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = cfg->reg_base;
+	struct it8xxx2_clock_control_cells *clk_cfg =
+		(struct it8xxx2_clock_control_cells *)(sub_system);
+
+	uint8_t offset = clk_cfg->offset;
+	uint8_t code = clk_cfg->bitmask;
+	uint8_t always = clk_cfg->always;
+
+	volatile uint8_t *reg =
+		(volatile uint8_t *)((uint32_t)ecpm_regs + (uint32_t)offset);
+
+	*reg |= (always | code);
+
+	return 0;
+}
+
+const uint32_t pll_reg_to_freq[8] = {MHZ(8), MHZ(16), MHZ(24), MHZ(32),
+				MHZ(48), MHZ(64), MHZ(72), MHZ(96)};
+/*
+ *  TODO: So far it only supports get PLL frequency but we'll add more
+ *  returned options in the future for available device nodes. In the future,
+ *  the related properties and nodes will be added in the it8xxx2.dtsi and
+ *  YAML file.
+ */
+static int it8xxx2_clock_control_get_rate(const struct device *dev,
+					  clock_control_subsys_t sub_system,
+					  uint32_t *rate)
+{
+	const struct clock_control_it8xxx2_config *const config = dev->config;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = config->reg_base;
+
+	int reg_val = ecpm_regs->PLLFREQR & PLLFREQ_MASK;
+
+	*rate = pll_reg_to_freq[reg_val];
+
+	return 0;
+}
+
+/*
+ *  TODO: So far it only supports set CPU frequency but we'll add more
+ *  returned options in the future for available device nodes. In the future,
+ *  the related properties and nodes will be added in the it8xxx2.dtsi and
+ *  YAML file.
+ */
+static int it8xxx2_clock_control_set_rate(const struct device *dev,
```

Comment:
@keith-zephyr 

Thanks for suggestions and guidance, sir.
I've updated the code to extend the flexibility in the future but so far there're some limitations when it comes to the options of the clock_control_get_rate() and clock_control_set_rate():

1. CPU Frequency:
The Clock Frequency of the IT8XXX2 processor can be determined according to the PLL Frequency and the divisor which is defined by the MCU Clock Frequency Select field in the register SCDCR0 of ECPM. Now we still keep this part and make the sub_system more complete so that it could probably support more features in the future but by now only the clock_control_get_rate() can have the options, the clock_control_set_rate() still provides the CPU frequency configuration only, and the details are described in the next two paragraphs. 

2. PLL Frequency:
We strongly suggest that the PLL frequency NOT be configured in real-time for the PLL frequency change requires some specific flows including making the EC enter the sleep mode, that is, during the period when EC be in sleep mode, the system could be unpredictable because of a millisecond-level latency. Therefore, we only configure the PLL during the EC initialization in the very beginning, and that's why we made it to be executed in the RAM because we just want it to be faster in the initialization.

`#define __soc_ram_code __attribute__((section(".__ram_code")))`

3. I2C/SMBus Source Clock Frequency:
We also strongly suggest that the I2C/SMBus Source Clock Frequency NOT be configured in real-time.

So far the it8xxx2_clock_control_set_rate() only supports the MCU clock configuration but the source clock frequency of the I2C/SMBus. For there are further configurations in I2C/SMBus registers which would put the source clock frequency into division again accordingly, so it's worth noticing that the source clock frequency of the I2C/SMBus is different from the final Speed/Frequency of the I2C/SMBus.

The final Sspeed/Frequency can only be obtained through related I2C/SMBus registers. Besides, the change of the I2C/SMBus Source Clock Frequency could lead that the frequency won't match the definitions in the register SCLKTS_A, that is, the Speed/frequency will be wrong in such a scenario. Hence, we will still not let the configurations on I2C/SMBus Source ClockFrequency be available in options.

To sum up, the it8xxx2_clock_control_set_rate() is extended to have the options in the sub_system but unfortunately it only supports the CPU clock frequency adjustment in realtime so far. However, the modifications about the sub_system are made for the possible extensions in the future.